### PR TITLE
feature(security): Adds functions to create and validate HMAC tokens

### LINKED ISSUE #7824: Make site secret available through public api
A few plugins need to use the site secret. Currently the function `get_site_secret` is access private. Is that on purpose? Do we need a new publicly available `elgg_get_site_secret` or a class function `$site->getSecret()`

### PR DESCRIPTION
**Pull request by mrclay**
This also has getHmac() use a binary encoding of the site key instead of the Base64 or hex encodings.

Fixes #7824

### CODE REVIEW

**Comment by jdalsem** [comment]
Does not really fix #7824 as site secret is still not publicly through API, but this PR will reduce the need to have it available a bit. Don't know if it will cover all usecases.

File: engine/tests/phpunit/ElggCryptoTest.php [hunk 1]
```diff
@@ -41,4 +45,31 @@ function provider() {
...
+
+	function testCanGenerateMacs() {
+		$crypto = new ElggCrypto();
+		$key = 'a very bad key';
+
+		$mac = $crypto->getHmac(1234, $key);
+		$this->assertTrue((bool)preg_match('~^[a-zA-Z0-9\-_]{30,}$~', $mac));
+
+		$mac = $crypto->getHmac([1234, 'foo'], $key);
+		$this->assertTrue((bool)preg_match('~^[a-zA-Z0-9\-_]{30,}$~', $mac));
```
Comment (ewinslow) [inline engine/tests/phpunit/ElggCryptoTest.php]: What is this magic regex?

**Comment by ewinslow** [comment]
Looks cool. Bummer about the new global method but whatevs I guess.

File: engine/classes/Elgg/ActionsService.php [hunk 2]
```diff
@@ -264,10 +264,10 @@ public function generateActionToken($timestamp) {
 		$site_secret = _elgg_services()->siteSecret->get();
```
Comment (jdalsem) [inline engine/classes/Elgg/ActionsService.php]: i believe you can drop this, as you are now defaulting

File: engine/lib/actions.php [hunk 3]
```diff
@@ -98,6 +98,38 @@ function elgg_unregister_action($action) {
...
+ * @return bool
+ * @since 1.10
+ * @see elgg_generate_mac
+ */
+function elgg_validate_mac($mac, $data, $algo = 'sha256') {
+	$crypto = _elgg_services()->crypto;
+	return $crypto->areEqual($mac, $crypto->getHmac($data, '', $algo));
+}
+
+/**
```
Comment (jdalsem) [inline engine/lib/actions.php]: do these functions really need to be in actions? I don't know a really good location either

**Comment by jdalsem** [comment]
> Don't know if it will cover all usecases.

i check with all our own usecases and it covers them all. So really happy to see this in core

File: docs/guides/actions.rst [hunk 4]
```diff
@@ -425,3 +425,30 @@ You can also access the tokens from javascript:
 These are refreshed periodically so should always be up-to-date.
 
 
+Security Tokens
+===============
+If you need to pass data to a user and trust that the user has returned it to you unaltered, you should use
```
Comment (juho-jaakkola) [inline docs/guides/actions.rst]: I had to read this twice to understand it. Could this be explained better?

**Comment by mrclay** [comment]
I've tried to address all these. Note that this will invalidate any unused tokens such as email validations/invitations. It uses a binary form of the site key instead of the hex/b64 encodings and also includes separators in the data.

I walked back the idea of auto-serializing data passed in because I was uncomfortable with having HMAC not match other implementations, but I'm still a bit torn on this. If someone fails to use a good separator, an attacker can move characters between variables: E.g. using `$data = $a . $b` means if you have a MAC for "123" and "hello", you can re-use that MAC for "12" and "3hello". That's not good.

@ewinslow Re global functions: https://community.elgg.org/discussion/view/2062019/ideas-for-exposing-elgg-oo-services

**Comment by beck24** [comment]
LGTM

File: engine/classes/Elgg/Database/SiteSecret.php [hunk 5]
```diff
@@ -36,14 +36,32 @@ function init() {
...
+	function get($raw = false) {
 		$secret = _elgg_services()->datalist->get('__site_secret__');
 		if (!$secret) {
 			$secret = init_site_secret();
 		}
+
+		if ($raw) {
+			// try to return binary key
+			if ($secret[0] === 'z') {
+				// new keys are "z" + base64URL
```
Comment (juho-jaakkola) [inline engine/classes/Elgg/Database/SiteSecret.php]: Why z? Might be best to document.

**Comment by mrclay** [comment]
I'm thinking of an API like:

```php
// HMAC with direct set of data
$token = elgg_build_hmac()->setData("$foo|$bar")->getToken();

$is_valid = elgg_build_hmac()->setData("$foo|$bar")->matchesToken($token);

// preferred usage: handles separation for you
$token = elgg_build_hmac()
    ->addData($foo)
    ->addData($bar)
    ->getToken();
```

**Comment by mrclay** [comment]
I reworked this so that it can accept an array of strings as HMAC data, and spits out an object instead of breaking generation/validation into two functions.

**Comment by juho-jaakkola** [comment]
LGTM

**Comment by hypeJunction** [comment]
I am unable to get this to work. For whatever reasons tokens do not match. Using hash_hmac() for now.

**Comment by mrclay** [comment]
@hypeJunction did you make sure the types match? http://learn.elgg.org/en/1.11/guides/actions.html#security-tokens

**Comment by hypeJunction** [comment]
I did.

**Comment by hypeJunction** [comment]
I am not sure if this affects the matching alrgorithm, but link was generated with elgg_build_hmac() and validation was being done without engine boot using Hmac component.

**Comment by mrclay** [comment]
Ah, few things: server key is decoded from b64/hex to binary before use, and the hash created is b64URL encoded. I should've made all those steps easily accessible without boot.

**Comment by mrclay** [comment]
At the very least I'll add that to docs for this function

**Pull request merged**